File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* core(int l, int r, std::vector<int>& nums){
        if(l > r) return nullptr;
        int mid = (l + r) / 2;
        TreeNode* root = new TreeNode(nums[mid]); //use the middle num as the root value.
        root->left = core(l, mid-1, nums); // create left subtree and append the root node.
        root->right = core(mid+1, r, nums); // create right subtree and append the root node.
        return root;
    }
public:
    TreeNode* sortedArrayToBST(std::vector<int>& nums) {
        return core(0, nums.size()-1, nums);
    }
};

File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private TreeNode core(int l, int r, int[] nums){
        if(l > r) return null;
        int mid = (l + r) / 2;
        TreeNode root = new TreeNode(nums[mid]); //use the middle num as the root value.
        root.left = core(l, mid-1, nums); // create left subtree and append the root node.
        root.right = core(mid+1, r, nums); // create right subtree and append the root node.
        return root;
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        return core(0, nums.length-1, nums);
    }
}
